File: gsonpath-compiler/src/main/java/gsonpath/model/GsonObject.java
package gsonpath.model;

import com.google.common.base.Objects;

import java.util.LinkedHashMap;

public class GsonObject {
    private final LinkedHashMap<String, Object> fieldMap;

    public GsonObject() {
        fieldMap = new LinkedHashMap<>();
    }

    public GsonObject addObject(String branchName, GsonObject gsonObject) {
        fieldMap.put(branchName, gsonObject);
        return gsonObject;
    }

    public GsonField addField(String branchName, GsonField field) throws IllegalArgumentException {
        if (containsKey(branchName)) {
            throw new IllegalArgumentException("Value already exists");
        }
        fieldMap.put(branchName, field);
        return field;
    }

    public int size() {
        return fieldMap.size();
    }

    public java.util.Set<String> keySet() {
        return fieldMap.keySet();
    }

    public Object get(String key) {
        return fieldMap.get(key);
    }

    public boolean containsKey(String key) {
        return fieldMap.containsKey(key);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        GsonObject gsonObject = (GsonObject) o;
        return Objects.equal(fieldMap, gsonObject.fieldMap);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(fieldMap);
    }

    @Override
    public String toString() {
        return "GsonObject: " + fieldMap;
    }
}

File: gsonpath-compiler/src/main/java/gsonpath/model/GsonObject.kt
package gsonpath.model

import com.google.common.base.Objects

import java.util.LinkedHashMap

class GsonObject {
    private val fieldMap: LinkedHashMap<String, Any>

    init {
        fieldMap = LinkedHashMap<String, Any>()
    }

    fun addObject(branchName: String, gsonObject: GsonObject): GsonObject {
        fieldMap[branchName] = gsonObject
        return gsonObject
    }

    @Throws(IllegalArgumentException::class)
    fun addField(branchName: String, field: GsonField): GsonField {
        if (containsKey(branchName)) {
            throw IllegalArgumentException("Value already exists")
        }
        fieldMap[branchName] = field
        return field
    }

    fun size(): Int {
        return fieldMap.size
    }

    fun keySet(): Set<String> {
        return fieldMap.keys
    }

    operator fun get(key: String): Any? {
        return fieldMap[key]
    }

    fun containsKey(key: String): Boolean {
        return fieldMap.containsKey(key)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val gsonObject = other as GsonObject?
        return Objects.equal(fieldMap, gsonObject!!.fieldMap)
    }

    override fun hashCode(): Int {
        return Objects.hashCode(fieldMap)
    }

    override fun toString(): String {
        return "GsonObject: " + fieldMap
    }
}
